examples/throughput/subscriber.cpp

```diff
@@ -0,0 +1,327 @@
+                  << " samples/s, " << static_cast<double>(total_bytes - prev_bytes) / (deltaTime*BYTES_PER_SEC_TO_MEGABITS_PER_SEC) << " Mbit/s, with "
+                  << outOfOrder << " samples received out of order.\n" << std::flush;
+        cycles++;
+        prev_time = time_now;
+        prev_bytes = total_bytes;
+        prev_samples = total_samples;
+      }
+    }
+  }
+
+  /* Output totals and averages */
+  auto deltaTime = static_cast<double>((std::chrono::steady_clock::now()-startTime).count())/1e9; //always 1e9?
+  std::cout << "\n" << subprefix << "Total received: " << total_samples << " samples, " << total_bytes << " bytes.\n";
+  std::cout << subprefix << "Out of order: " << outOfOrder << " samples.\n";
+  std::cout << subprefix << std::setprecision(4) << "Average transfer rate: " << static_cast<double>(total_samples) / deltaTime << " samples/s, "
+            << static_cast<double>(total_bytes) / (deltaTime*BYTES_PER_SEC_TO_MEGABITS_PER_SEC)  << " Mbit/s.\n" << std::flush;
+}
+
+static void sigint (int sig)
+{
+  (void) sig;
+  done = true;
+}
+
+template<typename T>
+bool wait_for_writer(dds::sub::DataReader<T> &reader)
+{
+  std::cout << "\n" << subprefix << "Waiting for a writer ...\n" << std::flush;
+
+  dds::core::cond::StatusCondition sc = dds::core::cond::StatusCondition(reader);
+  sc.enabled_statuses(dds::core::status::StatusMask::subscription_matched());
+
+  dds::core::cond::WaitSet waitset;
+  waitset.attach_condition(sc);
+
+  dds::core::cond::WaitSet::ConditionSeq conditions =
+    waitset.wait(dds::core::Duration::from_secs(30));
+
+  for (const auto & c:conditions)
+  {
+    if (c == sc)
+      return true;
+  }
+
+  std::cout << subprefix << "Did not discover a writer.\n" << std::flush;
+
+  return false;
+}
+
+template<typename T>
+class ThroughputListener: public dds::sub::DataReaderListener<T>
+{
+  public:
+  /*implementation of virtual functions*/
+
+  /*only on_data_available does anything*/
+  void on_data_available(dds::sub::DataReader<T>& rd) {
+    (void)do_take(rd);
+  }
+
+  /*all others are just dummies*/
+  void on_requested_deadline_missed(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::RequestedDeadlineMissedStatus&) { }
+
+  void on_requested_incompatible_qos(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::RequestedIncompatibleQosStatus&) { }
+
+  void on_sample_rejected(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::SampleRejectedStatus&) { }
+
+  void on_liveliness_changed(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::LivelinessChangedStatus&) { }
+
+  void on_subscription_matched(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::SubscriptionMatchedStatus&) { }
+
+  void on_sample_lost(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::SampleLostStatus&) { }
+};
```

Comment:
Instead of extending from the DataReaderListener, where all methods need to be overridden, why not extend from NoOpDataReaderListener, where only the one that you are going to implement needs to be overridden?

Reply:
Good suggestion.